go/pkg/reader/reader.go

```diff
@@ -209,33 +209,34 @@ func (cfs *compressedSeeker) Read(p []byte) (int, error) {
 	// to plan" error. EOF implies our uncompressed input data has finished reading,
 	// but it doesn't mean that we've finished preparing/returning our compressed data.
 	// We only want to return EOF to the user when the _compressed_ data is finished reading.
+	var finalErr error
 	if errR != nil && errR != io.EOF {
-		cfs.err = errR
+		finalErr = errR
 	} else if errW != nil {
-		cfs.err = errW
+		finalErr = errW
 	}
 
 	// When the buffer ends, or in case of _any_ error, we compress all the bytes we
 	// had available. The encoder requires a Close call to finish writing compressed
 	// data smaller than zstd's window size.
 	var errC error
-	if cfs.err != nil || errR == io.EOF {
+	if finalErr != nil || errR == io.EOF {
 		errC = cfs.encdW.Encoder.Close()
 		encoders.Put(cfs.encdW)
 		cfs.encdW = nil
```

Comment:
does the test exercise this code path? on the second time through it seems `cfs.encdW` will be nil and the call on line 223 will NPE?

before this change, the error caching logic prevented this block from executing a second time, but now it can.

Reply:
Good question, I think it's actually fine because `cfs.encdW != nil` in the `for` condition will evaluate to false and cause the loop to not run at all, so `finalErr` and `errR` will both be nil when we get here and the `if` won't be entered. The `cfs.buf.Read()` call after this should then correctly return `io.EOF`.

I added an assertion to cover this code path by calling `Read()` immediately after the `io.ReadAll()` and checking that it returns `io.EOF`.

It took a bit of inspection for me to figure that out though, so I'm open to suggestions for making the code more obviously correct.